Sources/PubNub/DependencyContainer/DependencyContainer.swift

```diff
@@ -0,0 +1,263 @@
+    get {
+      if let existingValue = values[ObjectIdentifier(key)] {
+        if let existingValue = existingValue as? K.Value {
+          return existingValue
+        } else {
+          preconditionFailure("Cannot resolve value for \(key)")
+        }
+      }
+      let value = key.value(from: self)
+      values[ObjectIdentifier(key)] = value
+      return value
+    } set {
+      values[ObjectIdentifier(key)] = newValue
+    }
+  }
+
+  @discardableResult
+  func register<K: DependencyKey>(value: K.Value?, forKey key: K.Type) -> DependencyContainer {
+    if let value {
+      values[ObjectIdentifier(key)] = value
+    }
+    return self
+  }
+}
+
+typealias SubscribeEngine = EventEngine<any SubscribeState, Subscribe.Event, Subscribe.Invocation, Subscribe.Dependencies>
+typealias PresenceEngine = EventEngine<any PresenceState, Presence.Event, Presence.Invocation, Presence.Dependencies>
+
+extension DependencyContainer {
+  var configuration: PubNubConfiguration {
+    self[PubNubConfigurationDependencyKey.self]
+  }
+
+  var instanceID: UUID {
+    self[PubNubInstanceIDDependencyKey.self]
+  }
+
+  var fileURLSession: URLSessionReplaceable {
+    self[FileURLSessionDependencyKey.self]
+  }
+
+  var subscriptionSession: SubscriptionSession {
+    self[SubscriptionSessionDependencyKey.self]
+  }
+
+  var presenceStateContainer: PubNubPresenceStateContainer {
+    self[PresenceStateContainerDependencyKey.self]
+  }
+
+  var defaultHTTPSession: SessionReplaceable {
+    resolveSession(
+      session: self[DefaultHTTPSessionDependencyKey.self],
+      with: [automaticRetry].compactMap { $0 }
+    )
+  }
+
+  fileprivate var httpSubscribeSession: SessionReplaceable {
+    resolveSession(
+      session: self[HTTPSubscribeSessionDependencyKey.self],
+      with: [instanceIDOperator].compactMap { $0 }
+    )
+  }
+
+  fileprivate var httpPresenceSession: SessionReplaceable {
+    resolveSession(
+      session: self[HTTPPresenceSessionDependencyKey.self],
+      with: [instanceIDOperator].compactMap { $0 }
+    )
+  }
+
+  fileprivate var automaticRetry: RequestOperator? {
+    configuration.automaticRetry
+  }
+
+  fileprivate var instanceIDOperator: RequestOperator? {
+    configuration.useInstanceId ? InstanceIdOperator(instanceID: instanceID.uuidString) : nil
+  }
+
+  fileprivate var httpSubscribeSessionQueue: DispatchQueue {
+    self[HTTPSubscribeSessionQueueDependencyKey.self]
+  }
+
+  fileprivate var subscribeEngine: SubscribeEngine {
+    self[SubscribeEventEngineDependencyKey.self]
+  }
+
+  fileprivate var presenceEngine: PresenceEngine {
+    self[PresenceEventEngineDependencyKey.self]
+  }
+}
+
+private extension DependencyContainer {
+  func resolveSession(session: SessionReplaceable, with operators: [RequestOperator?]) -> SessionReplaceable {
+    session.defaultRequestOperator == nil ? session.usingDefault(requestOperator: MultiplexRequestOperator(
+      operators: operators.compactMap { $0 }
```

Comment:
This is a private extension and can be sure of what we are passing, do we really need to map values again? We did it once when called `resolveSession`. 
Can we leave `compactMap` only in `resolveSession` and just pass arrays from `defaultHTTPSession`, `httpSubscribeSession`, and `httpPresenceSession`?

Reply:
Indeed, you're right. I will send you the updated version shortly.